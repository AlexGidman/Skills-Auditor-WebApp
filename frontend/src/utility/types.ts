export const SYSTEM_ROLES = ["StaffUser", "Manager", "Admin"];
export const STAFF_USER = SYSTEM_ROLES[0];
export const MANAGER_SR = SYSTEM_ROLES[1];
export const ADMIN = SYSTEM_ROLES[2];

export const JOB_ROLES = ["Mid-Level Developer", "Senior Developer", "Manager"];
export const MIDLEVEL_DEVELOPER = JOB_ROLES[0];
export const SENIOR_DEVELOPER = JOB_ROLES[1];
export const MANAGER_JR = JOB_ROLES[2];

export const TOKEN_COOKIE = "token";

// Week in seconds
export const TOKEN_LIFE = 60 * 60 * 24 * 7;

export const SKILL_LEVELS = ["None", "Basic", "Intermediate", "High", "Expert"];
export const NONE = SKILL_LEVELS[0];
export const BASIC = SKILL_LEVELS[1];
export const INTERMEDIATE = SKILL_LEVELS[2];
export const HIGH = SKILL_LEVELS[3];
export const EXPERT = SKILL_LEVELS[4];

export type User = {
    firstName: string;
    lastName: string;
    email: string;
    password: string;
};
