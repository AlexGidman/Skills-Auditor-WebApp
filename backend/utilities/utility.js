const argon2 = require("argon2");
const jwt = require("jsonwebtoken");
const { user, directReport } = require("../models");
const constants = require("../utilities/constants");

function formatErrorResponse(res, code, error) {
    console.warn(error);
    const message =
        error.name && error.name.includes("Sequelize") ? "Database error" : error.message;

    const err = {
        error: {
            status: code,
            message: message,
        },
    };
    return res.status(code).send(err);
}

function notValidUser(user) {
    if (
        !user.forename ||
        !user.surname ||
        !user.email ||
        !user.password ||
        !user.system_role ||
        !user.job_role
    ) {
        return true;
    }
    return false;
}

async function checkRegisteredRelationship(model, predicate) {
    const duplicate = await model.findOne(predicate);
    if (duplicate) {
        throw new Error("Manager assignment error: A relationship has already been made");
    }
}

async function checkForDuplicateEntry(model, predicate) {
    const duplicate = await model.findOne(predicate);
    if (duplicate) {
        throw new Error("Duplicate Entry");
    }
}

function changeObjectItemValue(payload, item, value) {
    if (Array.isArray(payload)) {
        for (let i = 0; i < payload.length; i++) {
            payload[i][item] = value;
        }
        return payload;
    } else {
        payload[item] = value;
        return payload;
    }
}

function isValidItem(arrayOfItems, item) {
    if (arrayOfItems) {
        for (let i = 0; i < arrayOfItems.length; i++) {
            if (item === arrayOfItems[i]) {
                return true;
            }
        }
    }
    return false;
}

function stringIsValidLength(item, min, max) {
    const lengthOfItem = item.length;
    return lengthOfItem >= min && lengthOfItem <= max;
}

const hashingOptions = { type: argon2.argon2id, memoryCost: 15360 }; // defaults - parallelism: 1, timeCost: 3 (iterations)

async function getPasswordHash(passwordText) {
    return await argon2.hash(passwordText, hashingOptions);
}

async function passwordsMatch(hash, passwordText) {
    return await argon2.verify(hash, passwordText, hashingOptions);
}

function generateAccessToken(id) {
    return jwt.sign({ id }, process.env.TOKEN_SECRET, { expiresIn: constants.JWT_EXPIRES_IN });
}

function checkToken(req, res, action) {
    const token = extractToken(req);
    if (!token || !constants.LOGIN_USER_ID) {
        return formatErrorResponse(res, 401, "Authorisation required");
    }

    jwt.verify(token, process.env.TOKEN_SECRET, (error, decoded) => {
        if (error) {
            return formatErrorResponse(res, 403, "Invalid Credentials");
        }
        req.tokenData = decoded;

        // Additional check token userID matched LOGIN_USER_ID so that our current checks are validated
        if (decoded.id != constants.LOGIN_USER_ID) {
            return formatErrorResponse(res, 403, "No Access");
        }
        res.locals.userId = decoded.id;

        action();
    });
}

function extractToken(req) {
    const authHeader = req.headers["authorization"];
    return authHeader && authHeader.split(" ")[1];
}

async function actionIsPermittedBySystemRole(userId, systemRoleArray) {
    try {
        const requestingUser = await user.findByPk(userId);
        if (!requestingUser) return false;
        for (const permittedSystemRole of systemRoleArray) {
            if (requestingUser.system_role === permittedSystemRole) {
                return true;
            }
        }
        return false;
    } catch (error) {
        return false;
    }
}

async function isUserOrDirectReportOfUser(requestingUserId, actionableUserId) {
    if (requestingUserId.toString() === actionableUserId.toString()) return true;
    try {
        const requestingUserDirectReports = await directReport.findAll({
            where: { user_id: requestingUserId },
        });
        if (!requestingUserDirectReports || !requestingUserDirectReports.length) return false;
        for (const report of requestingUserDirectReports) {
            if (report.report_id.toString() === actionableUserId.toString()) {
                return true;
            }
        }
        return false;
    } catch (error) {
        return false;
    }
}

module.exports = {
    formatErrorResponse,
    notValidUser,
    checkRegisteredRelationship,
    checkForDuplicateEntry,
    getPasswordHash,
    passwordsMatch,
    changeObjectItemValue,
    isValidItem,
    stringIsValidLength,
    generateAccessToken,
    checkToken,
    extractToken,
    actionIsPermittedBySystemRole,
    isUserOrDirectReportOfUser,
};
