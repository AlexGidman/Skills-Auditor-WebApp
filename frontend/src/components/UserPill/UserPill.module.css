.component {
}

.statusBadge {
    border: solid 3px;
    border-radius: var(--border-rad);
    padding: 0 var(--s1);
    vertical-align: middle;
    margin-right: calc(var(--s1) / 2);
}

.admin {
    composes: statusBadge;
    color: gold;
}

.manager {
    composes: statusBadge;
    color: var(--c-white);
}

.pill {
    border: solid 3px;
    border-radius: var(--s2);
    padding: var(--s1);
    max-width: var(--s10);
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: small;
    color: var(--c-white);
    background-color: var(--c-secondary);
}

@media (max-width: 900px) {
    .statusBadge {
        display: none;
    }
}
