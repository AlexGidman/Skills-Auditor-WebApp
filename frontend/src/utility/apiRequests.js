import axios from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders, id) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders, name) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (authHeaders, id, name) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders, id) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (authHeaders, name, category_id) => {
    const data = {
        name,
        category_id,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (authHeaders, id, name, categoryId) => {
    const data = {
        id,
        name,
        category_id: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders, id) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders,
    id,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        id,
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders, id) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (authHeaders, userId, reportId) => {
    const data = {
        user_id: userId,
        report_id: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders,
    user_id,
    skill_id,
    skill_level,
    notes,
    expiry_date,
) => {
    const data = {
        user_id,
        skill_id,
        skill_level,
        notes,
        expiry_date,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (authHeaders, id, skill_level, notes) => {
    const data = {
        id,
        skill_level,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email, password) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
