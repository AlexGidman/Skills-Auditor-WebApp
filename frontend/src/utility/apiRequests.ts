import axios, { AxiosRequestHeaders } from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders: AxiosRequestHeaders, name: string) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (
    authHeaders: AxiosRequestHeaders,
    name: string,
    categoryId: string,
) => {
    const data = {
        name,
        categoryId,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
    categoryId: string,
) => {
    const data = {
        id,
        name,
        categoryId: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        id,
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders: AxiosRequestHeaders,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    reportId: string,
) => {
    const data = {
        userId: userId,
        reportId: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    skillId: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
    expiryDate: Date,
) => {
    const data = {
        userId,
        skillId,
        skillLevel,
        notes,
        expiryDate,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
) => {
    const data = {
        id,
        skillLevel,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email: string, password: string) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
