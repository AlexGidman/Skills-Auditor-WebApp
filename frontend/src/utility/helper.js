import { useEffect, useState } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import {
    ADMIN,
    BASIC,
    EXPERT,
    HIGH,
    NONE,
    INTERMEDIATE,
    MANAGER_SR,
    SYSTEM_ROLES,
    TOKEN_COOKIE,
} from "./types";
import { useCookies } from "react-cookie";

export const useLogout = () => {
    const [, , removeCookie] = useCookies(TOKEN_COOKIE);
    const navigate = useNavigate();
    const [logout, setLogout] = useState(false);

    useEffect(() => {
        if (logout) {
            setLogout(false);
            removeCookie(TOKEN_COOKIE);
            navigate("/login");
        }
    }, [logout, navigate, removeCookie]);

    return () => setLogout(true);
};

export const hyphenate = (string) => {
    return string.trim().replace(" ", "-");
};

export const useAPI = (apiCallback, varArray = [], lazy = false, auth = true) => {
    const [cookies] = useCookies(TOKEN_COOKIE);
    const navigate = useNavigate();

    const [callAPI, setCallAPI] = useState(!lazy);
    const [data, setData] = useState(null);
    const [loading, setLoading] = useState(!lazy);
    const [error, setError] = useState(null);

    const callback = () => {
        setLoading(true);
        setCallAPI(true);
    };

    const authHeaders = {
        headers: {
            authorization: cookies[TOKEN_COOKIE],
        },
    };

    useEffect(() => {
        if (callAPI) {
            apiCallback
                .apply(null, auth ? [authHeaders, ...varArray] : [...varArray])
                .then((response) => {
                    setData(response.data);
                })
                .catch((error) => {
                    if (unauthorisedError(error)) {
                        navigate("/login");
                    } else {
                        setError(error.response.data.error);
                    }
                })
                .finally(() => {
                    setLoading(false);
                    setCallAPI(false);
                });
        }
    }, [apiCallback, auth, callAPI, navigate, varArray]);

    return { data, loading, error, callback };
};

export const getSelectOptionsFromArray = (valArray = [], labelArray = []) => {
    return valArray.length === labelArray.length
        ? valArray.map((val, i) => {
              return { value: val, label: labelArray[i] };
          })
        : valArray.map((val) => {
              return { value: val, label: val };
          });
};

export const getAssignableSystemRoleOptions = (currentUserSystemRole) => {
    const systemRolesOptions = getSelectOptionsFromArray(SYSTEM_ROLES);
    switch (currentUserSystemRole) {
        case ADMIN:
            return systemRolesOptions;
        case MANAGER_SR:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option;
            });
        default:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option.label !== MANAGER_SR && option;
            });
    }
};

export const isValidString = (val, min = 0, max = 100) => {
    return !!val && val.length >= min && val.length <= max;
};

export const isValidEmail = (val) => {
    return /\S+@\S+\.\S+/.test(val);
};

export const isValidUser = (user) => {
    return (
        !!user.firstName &&
        isValidString(user.firstName, 1) &&
        !!user.lastName &&
        isValidString(user.lastName, 1) &&
        !!user.email &&
        isValidEmail(user.email) &&
        (!!user.password ? isValidString(user.password, 10, 16) : true)
    );
};

export const useIsAdminOrManager = () => {
    const [currentUser, setShowToast] = useOutletContext();
    const navigate = useNavigate();

    useEffect(() => {
        if (
            currentUser &&
            currentUser.system_role !== ADMIN &&
            currentUser.system_role !== MANAGER_SR
        ) {
            setShowToast({ error: "Unauthorised route" });
            navigate("/");
        }
    });
};

const unauthorisedError = (error) => error.response.status === 401 || error.response.status === 403;

export const formatDate = (_date) => {
    if (!_date) return "N/A";
    let newDate = new Date(_date);
    newDate.toLocaleDateString();
    return newDate.toLocaleDateString();
};

export const formatSkillLevel = (_skillLevel) => {
    if (_skillLevel === 5) {
        return EXPERT;
    } else if (_skillLevel === 4) {
        return HIGH;
    } else if (_skillLevel === 3) {
        return INTERMEDIATE;
    } else if (_skillLevel === 2) {
        return BASIC;
    } else {
        return NONE;
    }
};

export const convertSkillLevelName = (_skillLevel) => {
    if (_skillLevel === "Expert") {
        return 5;
    } else if (_skillLevel === "High") {
        return 4;
    } else if (_skillLevel === "Intermediate") {
        return 3;
    } else if (_skillLevel === "Basic") {
        return 2;
    } else {
        return 1;
    }
};

export const expiryDateNullFormatter = (_expiryDate) => {
    if (_expiryDate === null) {
        return "N/A";
    } else {
        return _expiryDate;
    }
};
