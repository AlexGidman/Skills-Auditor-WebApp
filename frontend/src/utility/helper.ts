import { useEffect, useState } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import {
    ADMIN,
    BASIC,
    EXPERT,
    HIGH,
    NONE,
    INTERMEDIATE,
    MANAGER_SR,
    systemRoleS,
    TOKEN_COOKIE,
    User,
} from "./types";
import { useCookies } from "react-cookie";
import { AxiosResponse, AxiosError } from "axios";

export const useLogout = () => {
    const [, , removeCookie] = useCookies([TOKEN_COOKIE]); // TODO: Check this works!
    const navigate = useNavigate();
    const [logout, setLogout] = useState(false);

    useEffect(() => {
        if (logout) {
            setLogout(false);
            removeCookie(TOKEN_COOKIE);
            navigate("/login");
        }
    }, [logout, navigate, removeCookie]);

    return () => setLogout(true);
};

export const hyphenate = (str: string) => {
    return str.trim().replace(" ", "-");
};

export function useAPI<T>(
    apiCallback: Function,
    varArray: unknown[] = [],
    lazy = false,
    auth = true,
): { data: T | null; loading: boolean; error: Error | null; callback: CallableFunction } {
    const [cookies] = useCookies([TOKEN_COOKIE]); // TODO: Check this works!
    const navigate = useNavigate();

    const [callAPI, setCallAPI] = useState<boolean>(!lazy);
    const [data, setData] = useState<T | null>(null);
    const [loading, setLoading] = useState<boolean>(!lazy);
    const [error, setError] = useState<Error | null>(null);

    const callback = () => {
        setLoading(true);
        setCallAPI(true);
    };

    useEffect(() => {
        if (callAPI) {
            const authHeaders = {
                headers: {
                    authorization: cookies[TOKEN_COOKIE],
                },
            };

            apiCallback
                .apply(null, auth ? [authHeaders, ...varArray] : [...varArray])
                ?.then((response: AxiosResponse) => {
                    setData(response.data);
                })
                .catch((error: AxiosError<{ error: Error }>) => {
                    if (unauthorisedError(error)) {
                        navigate("/login");
                    } else {
                        setError(error?.response?.data?.error || null);
                    }
                })
                .finally(() => {
                    setLoading(false);
                    setCallAPI(false);
                });
        }
    }, [apiCallback, auth, callAPI, cookies, navigate, varArray]);

    return { data, loading, error, callback };
}

export const getSelectOptionsFromArray = (valArray: string[] = [], labelArray: string[] = []) => {
    return valArray.length === labelArray.length
        ? valArray.map((val, i) => {
              return { value: val, label: labelArray[i] };
          })
        : valArray.map((val) => {
              return { value: val, label: val };
          });
};

export const getAssignableSystemRoleOptions = (currentUserSystemRole: string) => {
    // TODO implement proper type for system role
    const systemRolesOptions = getSelectOptionsFromArray(systemRoleS);
    switch (currentUserSystemRole) {
        case ADMIN:
            return systemRolesOptions;
        case MANAGER_SR:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option;
            });
        default:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option.label !== MANAGER_SR && option;
            });
    }
};

export const isValidString = (val: string, min = 0, max = 100) => {
    return !!val && val.length >= min && val.length <= max;
};

export const isValidEmail = (val: string) => {
    return /\S+@\S+\.\S+/.test(val);
};

export const isValidUser = (user: User) => {
    return (
        !!user.firstName &&
        isValidString(user.firstName, 1) &&
        !!user.lastName &&
        isValidString(user.lastName, 1) &&
        !!user.email &&
        isValidEmail(user.email) &&
        (!!user.password ? isValidString(user.password, 10, 16) : true)
    );
};

export const useIsAdminOrManager = () => {
    // @ts-ignore TODO fix type for AppOutletContext here
    const [currentUser, setShowToast] = useOutletContext();
    const navigate = useNavigate();

    useEffect(() => {
        if (
            currentUser &&
            currentUser.systemRole !== ADMIN &&
            currentUser.systemRole !== MANAGER_SR
        ) {
            setShowToast({ error: "Unauthorised route" });
            navigate("/");
        }
    });
};

const unauthorisedError = (error: AxiosError): boolean => {
    return (
        !!!error?.response?.status ||
        error?.response?.status === 401 ||
        error?.response?.status === 403
    );
};

export const formatDate = (_date: string) => {
    if (!_date || _date === "0000-00-00") return "N/A";
    let newDate = new Date(_date);
    newDate.toLocaleDateString();
    return newDate.toLocaleDateString();
};

export const formatSkillLevel = (_skillLevel: string) => {
    // TODO implement proper type for skill level
    if (_skillLevel === "5") {
        return EXPERT;
    } else if (_skillLevel === "4") {
        return HIGH;
    } else if (_skillLevel === "3") {
        return INTERMEDIATE;
    } else if (_skillLevel === "2") {
        return BASIC;
    } else {
        return NONE;
    }
};

export const convertSkillLevelName = (_skillLevel: string) => {
    // TODO implement proper type for skill level
    if (_skillLevel === "Expert") {
        return "5";
    } else if (_skillLevel === "High") {
        return "4";
    } else if (_skillLevel === "Intermediate") {
        return "3";
    } else if (_skillLevel === "Basic") {
        return "2";
    } else {
        return "1";
    }
};
