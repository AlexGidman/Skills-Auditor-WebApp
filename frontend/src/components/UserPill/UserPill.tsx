import React from "react";
import cx from "classnames";
import styles from "./UserPill.module.css";
import { MANAGER_SR, STAFF_USER, ADMIN, SYSTEM_ROLES } from "../../utility/types";

interface Props extends React.ComponentProps<"div"> {
    email?: string;
    systemRole?: typeof SYSTEM_ROLES[number];
}

export const UserPill = ({
    email = "No user",
    systemRole = STAFF_USER,
    className,
    ...rest
}: Props) => {
    return (
        <div className={cx(styles.component, className)} {...rest}>
            {systemRole === MANAGER_SR && (
                <span
                    role="status"
                    aria-label="You are logged in as a manager"
                    className={styles.manager}
                >
                    MANAGER
                </span>
            )}
            {systemRole === ADMIN && (
                <span
                    role="status"
                    aria-label="You are logged in as an admin"
                    className={styles.admin}
                >
                    ADMIN
                </span>
            )}
            <span role="status" className={styles.pill} aria-label={`${email} is logged in`}>
                {email}
            </span>
        </div>
    );
};
