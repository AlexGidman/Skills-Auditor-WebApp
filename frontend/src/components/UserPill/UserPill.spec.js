import { render, screen } from "@testing-library/react";
import { ADMIN, MANAGER_SR, STAFF_USER } from "../../utility/types";
import { UserPill } from "./UserPill";

describe("UserPill", () => {
    it("renders correctly when passed manager props", async () => {
        const { container } = render(<UserPill email="test@email.com" systemRole={MANAGER_SR} />);

        expect(
            await screen.findByRole("status", { name: "test@email.com is logged in" }),
        ).toBeInTheDocument();
        expect(
            screen.getByRole("status", { name: "You are logged in as a manager" }),
        ).toBeInTheDocument();
        expect(
            screen.queryByRole("status", { name: "You are logged in as an admin" }),
        ).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
    it("renders correctly when passed admin props", async () => {
        const { container } = render(<UserPill email="test@email.com" systemRole={ADMIN} />);

        expect(
            await screen.findByRole("status", { name: "test@email.com is logged in" }),
        ).toBeInTheDocument();
        expect(
            screen.getByRole("status", { name: "You are logged in as an admin" }),
        ).toBeInTheDocument();
        expect(
            screen.queryByRole("status", { name: "You are logged in as a manager" }),
        ).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
    it("renders correctly when passed staff user props", async () => {
        const { container } = render(<UserPill email="test@email.com" systemRole={STAFF_USER} />);

        expect(
            await screen.findByRole("status", { name: "test@email.com is logged in" }),
        ).toBeInTheDocument();
        expect(
            screen.queryByRole("status", { name: "You are logged in as a manager" }),
        ).not.toBeInTheDocument();
        expect(
            screen.queryByRole("status", { name: "You are logged in as an admin" }),
        ).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
    it("renders correctly when passed no props", async () => {
        const { container } = render(<UserPill />);

        expect(
            await screen.findByRole("status", { name: "No user is logged in" }),
        ).toBeInTheDocument();
        expect(
            screen.queryByRole("status", { name: "You are logged in as a manager" }),
        ).not.toBeInTheDocument();
        expect(
            screen.queryByRole("status", { name: "You are logged in as an admin" }),
        ).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});
