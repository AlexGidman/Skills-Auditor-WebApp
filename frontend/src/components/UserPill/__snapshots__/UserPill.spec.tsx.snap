// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPill renders correctly when passed admin props 1`] = `
<div>
  <div
    class="component"
  >
    <span
      aria-label="You are logged in as an admin"
      class="admin"
      role="status"
    >
      ADMIN
    </span>
    <span
      aria-label="test@email.com is logged in"
      class="pill"
      role="status"
    >
      test@email.com
    </span>
  </div>
</div>
`;

exports[`UserPill renders correctly when passed manager props 1`] = `
<div>
  <div
    class="component"
  >
    <span
      aria-label="You are logged in as a manager"
      class="manager"
      role="status"
    >
      MANAGER
    </span>
    <span
      aria-label="test@email.com is logged in"
      class="pill"
      role="status"
    >
      test@email.com
    </span>
  </div>
</div>
`;

exports[`UserPill renders correctly when passed no props 1`] = `
<div>
  <div
    class="component"
  >
    <span
      aria-label="No user is logged in"
      class="pill"
      role="status"
    >
      No user
    </span>
  </div>
</div>
`;

exports[`UserPill renders correctly when passed staff user props 1`] = `
<div>
  <div
    class="component"
  >
    <span
      aria-label="test@email.com is logged in"
      class="pill"
      role="status"
    >
      test@email.com
    </span>
  </div>
</div>
`;
